service/matching/incoming_service_manager.go

```diff
@@ -0,0 +1,231 @@
+// The MIT License
+//
+// Copyright (c) 2024 Temporal Technologies Inc.  All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package matching
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"sync/atomic"
+
+	"github.com/google/uuid"
+
+	"go.temporal.io/server/api/matchingservice/v1"
+	persistencepb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/clock"
+	hlc "go.temporal.io/server/common/clock/hybrid_logical_clock"
+	"go.temporal.io/server/common/nexus"
+	p "go.temporal.io/server/common/persistence"
+)
+
+const (
+	// loadServicesPageSize is the page size to use when initially loading services from persistence
+	loadServicesPageSize = 100 // TODO: should be dynamic config?
+)
+
+type (
+	// incomingServiceManager manages cache and persistence access for Nexus incoming services.
+	// incomingServiceManager also contains a RWLock to enforce serial updates to prevent
+	// nexus_incoming_services table version conflicts.
+	//
+	// incomingServiceManager should only be used within matching service because it assumes
+	// that it is running on the matching node that owns the nexus_incoming_services table.
+	// There is no explicit listener for membership changes because table ownership changes
+	// will be detected by version conflicts and eventually settle through retries.
+	//
+	// Not to be confused with persistence.NexusServiceManager which is responsible for persistence-layer
+	// CRUD APIs for Nexus incoming services.
+	incomingServiceManager interface {
```

Comment:
Why do you need the interface?
Also the comment should be on the struct because it describes the implementation.

Reply:
It isn't completely necessary, but I like to have the interface definition and the comments together in one place because I think it is easier to read at a glance. I've noticed we have a similar pattern in other components, e.g. https://github.com/temporalio/temporal/blob/main/service/matching/task_queue_manager.go but I'm not attached to this format.